### ISSUE TITLE
design Object Model for PyInstaller

### ISSUE THREAD

**Issue by matysek** [pyinstaller/pyinstaller#1691]
* PyInstaller evolved over time and functionality was nod added in a consistent manner.
* Some functionality could be better expressed as classes.
* PyInstaller does not have well defined internal components and how they interact.
* We should more take advantage of advanced features of Python. E.g. inheritance, metaclasses, mixins, iterators, generators, context manager, ...

Output of this issue should be:
* high-level design of various internal PyInstaller components/objects and how they interact.
* list of hierarchy of Python modules and what code should live there.

Object module should be the basis for futher PyInstaller refactoring.

**Comment by codewarrior0** [comment]
`PyInstaller.depend.bindepend` has a `seen` global that has to be reset at the beginning of Analysis. This is a sign that the bindepend functionality should be refactored into a class, so the instance has the `seen` data and a new instance can be created for each Analysis.

**Comment by htgoebel** [comment]
I agree in the idea to use much more objects instead of "dump" data-types.

But IMHO we should not over-engineer here. PyInstaller is meant to build *one* application at a time. Thus global variables only matter for the test-suite. Eliminating them would most probably lead to passing around some "status" instances. So we should spend our time to more important topics.

**Comment by xoviat** [comment]
Brought about by #2337 and #2363, I think that modulegraph should be redesigned to:

1. Use an iterative rather than recursive mechanism.
2. Provide a hook api for evaluating code instructions; incorporate ctypes detection into this public api rather than having it in utils.
